File: mobile/modules/acs-meeting/ios/PcmBridge.swift
import Foundation

/// Downmix/resample PCM16 to 16 kHz mono ~20 ms blocks for ACS raw outgoing audio.
final class PcmBridge {
  private var dump = Data()
  private var dumpRate = 0
  private var dumpChannels = 0
  private var dumped = false
  private var pending = Data()
  private let dumpWav: Bool
  private var lastRmsLog = Date.distantPast

  init(dumpWav: Bool) {
    self.dumpWav = dumpWav
  }

  func ingest(pcm16Le: Data, sampleRate: Int, channels: Int) -> [Data] {
    if dumpWav, !dumped {
      dump.append(pcm16Le)
      dumpRate = sampleRate
      dumpChannels = channels
      if dump.count >= 30 * sampleRate * channels * 2 {
        finishDump()
      }
    }
    logLevel(pcm16Le, sampleRate: sampleRate, channels: channels)
    let mono = Self.toMono16k(pcm16Le, sampleRate: sampleRate, channels: channels)
    pending.append(mono)
    let frameBytes = 16000 * 2 / 50
    var frames: [Data] = []
    while pending.count >= frameBytes {
      frames.append(pending.prefix(frameBytes))
      pending.removeFirst(frameBytes)
    }
    return frames
  }

  func finishDump() {
    guard dumpWav, !dumped else { return }
    dumped = true
    let url = FileManager.default.temporaryDirectory.appendingPathComponent("acs-whep-p4.wav")
    try? Self.wav(pcm: dump, sampleRate: dumpRate, channels: dumpChannels).write(to: url)
    NSLog("ACS-SPIKE P4 wrote \(url.path) bytes=\(dump.count) rate=\(dumpRate) ch=\(dumpChannels)")
    dump.removeAll()
  }

  private func logLevel(_ pcm: Data, sampleRate: Int, channels: Int) {
    let now = Date()
    guard now.timeIntervalSince(lastRmsLog) >= 1 else { return }
    lastRmsLog = now
    var acc: Int64 = 0
    var count = 0
    pcm.withUnsafeBytes { raw in
      let samples = raw.bindMemory(to: Int16.self)
      for sample in samples {
        acc += Int64(abs(Int32(sample)))
        count += 1
      }
    }
    let mean = count == 0 ? 0 : acc / Int64(count)
    NSLog("ACS-SPIKE P4 pcm rate=\(sampleRate) ch=\(channels) meanAbs=\(mean) bytes=\(pcm.count)")
  }

  static func toMono16k(_ pcm16Le: Data, sampleRate: Int, channels: Int) -> Data {
    let samples = pcm16Le.withUnsafeBytes { Array($0.bindMemory(to: Int16.self)) }
    let mono: [Int16]
    if channels <= 1 {
      mono = samples
    } else {
      mono = stride(from: 0, to: samples.count, by: channels).map { index in
        var acc = 0
        for channel in 0 ..< channels where index + channel < samples.count {
          acc += Int(samples[index + channel])
        }
        return Int16(acc / channels)
      }
    }
    if sampleRate == 16000 {
      return mono.withUnsafeBufferPointer { Data(buffer: $0) }
    }
    let outCount = max(1, mono.count * 16000 / sampleRate)
    var resampled = [Int16](repeating: 0, count: outCount)
    for i in 0 ..< outCount {
      let src = min(mono.count - 1, i * sampleRate / 16000)
      resampled[i] = mono[src]
    }
    return resampled.withUnsafeBufferPointer { Data(buffer: $0) }
  }

  static func wav(pcm: Data, sampleRate: Int, channels: Int) -> Data {
    var header = Data()
    func put(_ value: UInt32) {
      var le = value.littleEndian
      header.append(Data(bytes: &le, count: 4))
    }
    func put16(_ value: UInt16) {
      var le = value.littleEndian
      header.append(Data(bytes: &le, count: 2))
    }
    header.append(contentsOf: [UInt8]("RIFF".utf8))
    put(UInt32(36 + pcm.count))
    header.append(contentsOf: [UInt8]("WAVE".utf8))
    header.append(contentsOf: [UInt8]("fmt ".utf8))
    put(16)
    put16(1)
    put16(UInt16(channels))
    put(UInt32(sampleRate))
    put(UInt32(sampleRate * channels * 2))
    put16(UInt16(channels * 2))
    put16(16)
    header.append(contentsOf: [UInt8]("data".utf8))
    put(UInt32(pcm.count))
    return header + pcm
  }
}

File: mobile/modules/acs-meeting/android/src/main/java/com/mentra/acsmeeting/audio/PcmBridge.kt
package com.mentra.acsmeeting.audio

import android.util.Base64
import android.util.Log
import java.io.ByteArrayOutputStream
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.abs

/** Downmix/resample PCM16 to 48 kHz mono ~20 ms blocks for ACS raw outgoing audio. */
class PcmBridge(
  private val dumpDir: File?,
  private val dumpWav: Boolean,
) {
  private var dump: ByteArrayOutputStream? = if (dumpWav) ByteArrayOutputStream() else null
  private var dumpRate = 0
  private var dumpChannels = 0
  private var dumped = false
  private val outgoing = ByteArrayOutputStream()
  private val resampler = PcmResampler(TARGET_RATE)
  private var lastRmsLogMs = 0L
  @Volatile var lastMeanAbs: Int = -1
    private set

  fun ingest(pcm16Le: ByteArray, sampleRate: Int, channels: Int): List<ByteArray> {
    if (dump != null && !dumped) {
      dump!!.write(pcm16Le)
      dumpRate = sampleRate
      dumpChannels = channels
      val needed = 30 * sampleRate * channels * 2
      if (dump!!.size() >= needed) finishDump()
    }
    logLevel(pcm16Le, sampleRate, channels)
    // Stateful: filter history and fractional phase carry across WebRTC's
    // 10 ms callbacks, so there is no seam every 480 samples.
    val mono = resampler.process(pcm16Le, sampleRate, channels)
    outgoing.write(mono)
    val frameBytes = TARGET_RATE * 2 / 50 // 20 ms
    val frames = mutableListOf<ByteArray>()
    val buf = outgoing.toByteArray()
    var offset = 0
    while (buf.size - offset >= frameBytes) {
      frames.add(buf.copyOfRange(offset, offset + frameBytes))
      offset += frameBytes
    }
    outgoing.reset()
    if (offset < buf.size) outgoing.write(buf, offset, buf.size - offset)
    return frames
  }

  /**
   * Drop the state that spans calls to [ingest]: the resampler's filter history and fractional
   * phase, and the partial 20 ms frame not yet emitted.
   *
   * What makes this necessary is that both are, by design, audio the wearer already spoke. A mute
   * that left them behind would emit them on the next unmute — a fraction of a word, from before
   * the wearer asked not to be heard. Serialization is the caller's job; see [AudioUplinkChain].
   */
  fun reset() {
    resampler.reset()
    outgoing.reset()
  }

  @Synchronized
  fun finishDump() {
    if (dumped) return
    val data = dump?.toByteArray() ?: return
    dumped = true
    dump = null
    val dir = dumpDir ?: return
    val file = File(dir, "acs-whep-p4.wav")
    file.writeBytes(wav(data, dumpRate, dumpChannels))
    Log.i(TAG, "P4 wrote ${file.absolutePath} bytes=${data.size} rate=$dumpRate ch=$dumpChannels")
  }

  private fun logLevel(pcm: ByteArray, sampleRate: Int, channels: Int) {
    var acc = 0L
    var n = 0
    var i = 0
    while (i + 1 < pcm.size) {
      val s = (pcm[i].toInt() and 0xff) or (pcm[i + 1].toInt() shl 8)
      val signed = if (s >= 0x8000) s - 0x10000 else s
      acc += abs(signed)
      n++
      i += 2
    }
    val mean = if (n == 0) 0 else acc / n
    lastMeanAbs = mean.toInt()
    val now = System.currentTimeMillis()
    if (now - lastRmsLogMs < 1000) return
    lastRmsLogMs = now
    val verdict = if (mean < SILENCE_MEAN_ABS) "SILENCE" else "voice"
    Log.i(TAG, "P4 pcm rate=$sampleRate ch=$channels meanAbs=$mean bytes=${pcm.size} $verdict")
  }

  companion object {
    private const val TAG = "ACS-SPIKE"
    const val TARGET_RATE = 48_000

    /** One-shot convenience (fresh filter state). Streams must use a [PcmResampler] instance. */
    fun toMono(pcm16Le: ByteArray, sampleRate: Int, channels: Int): ByteArray =
      PcmResampler(TARGET_RATE).process(pcm16Le, sampleRate, channels)

    fun wav(pcm: ByteArray, sampleRate: Int, channels: Int): ByteArray {
      val header = ByteBuffer.allocate(44).order(ByteOrder.LITTLE_ENDIAN)
      header.put("RIFF".toByteArray())
      header.putInt(36 + pcm.size)
      header.put("WAVE".toByteArray())
      header.put("fmt ".toByteArray())
      header.putInt(16)
      header.putShort(1)
      header.putShort(channels.toShort())
      header.putInt(sampleRate)
      header.putInt(sampleRate * channels * 2)
      header.putShort((channels * 2).toShort())
      header.putShort(16)
      header.put("data".toByteArray())
      header.putInt(pcm.size)
      return header.array() + pcm
    }

    fun encodeBase64(pcm: ByteArray): String = Base64.encodeToString(pcm, Base64.NO_WRAP)

    /**
     * 16-bit speech sits in the hundreds to thousands. Below this is the LC3/ADC noise floor
     * Teams hears as mute even when ACS is unmuted and the pacer is RUNNING.
     */
    const val SILENCE_MEAN_ABS = 50
  }
}
